### PR TITLE
reorganize cmake files(issue #113)

### PR DESCRIPTION
**Pull request by noskill**
It's my attempt at reorganizing cmake files https://github.com/opencog/cogutil/issues/113
1) move Find*.cmake files from atomspace to cogutil for installation
2) generate CogUtilConfig.cmake so that FindCogUtil.cmake  is not necessary

### CODE REVIEW

**Comment by linas** [comment]
Some fixes are needed; the comments are similar to those of  opencog/atomspace#1832 

* Don't remove the COPYING file, it pertains to the files in this directory.

* The valgrind, vagrant, buildbot, conf and spec files have nothing to do with cmake.  They belong in some other, appropriately-named directory.

**Comment by noskill** [comment]
updated

**Comment by ngeiswei** [comment]
It seems to make sense (although,with the exception of `cmake/CMakeLists.txt`, I didn't carefully review it).

I suppose one of us need to test it on a fresh system before merging (I'll test it early next week if none of us has done it by then).

**Comment by vsbogd** [comment]
On my machine ```make install`` copied files CogUtilConfig.cmake  CogUtilConfigVersion.cmake into root directory. I have passed installation path to ```cmake``` manually via CMAKE_INSTALL_PREFIX:

    $ cmake -DCMAKE_INSTALL_PREFIX=$OPENCOG/test/build $OPENCOG/test/cogutils
    $ make install
    $ ls $OPENCOG/test/build
    CogUtilConfig.cmake  CogUtilConfigVersion.cmake  include  lib  share

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -3,3 +3,53 @@ CONFIGURE_FILE(${CMAKE_SOURCE_DIR}/lib/opencog.conf
...
+  "${CMAKE_CURRENT_BINARY_DIR}/CogUtilConfigVersion.cmake"
+  VERSION ${SEMANTIC_VERSION}
+  COMPATIBILITY SameMajorVersion
+)
+
+INSTALL(FILES
+       ${CMAKE_CURRENT_BINARY_DIR}/CogUtilConfigVersion.cmake
+       ${CMAKE_CURRENT_BINARY_DIR}/CogUtilConfig.cmake
+       DESTINATION
+       ${CMAKE_INSTALL_PREFIX})
```
Comment (vsbogd) [inline cmake/CMakeLists.txt]: Should be ${LIB_INSTALL_DIR}/CogUtil/cmake probably

Comment (noskill) [reply cmake/CMakeLists.txt]: Thanks!

**Comment by noskill** [comment]
@vsbogd Thanks, i fixed include path, it should work now with custom prefix.

**Comment by vsbogd** [comment]
Checked again, CogUtil buils succesfully. I also can build atomspace but cannot build atomspace unit tests. See linkage error:
```
[ 17%] Linking CXX executable DefineLinkUTest
CMakeFiles/TVMergeUTest.dir/TVMergeUTest.cpp.o: In function `TVMergeUTest::testMerge()':
/home/vital/projects/opencog/test/atomspace-build/tests/truthvalue/../../../atomspace/tests/truthvalue/TVMergeUTest.cxxtest:96: undefined reference to `opencog::logger()'
...
```

**Comment by vsbogd** [comment]
CogUtil build successful, unit tests passed. Atomspace build successful, unit tests failed, but this is equal to baseline (opencog/atomspace master branch):
```
96% tests passed, 4 tests failed out of 113

Total Test time (real) = 233.34 sec

The following tests FAILED:
	 49 - MultiAtomSpaceUTest (Failed)
	 53 - VectorAPIUTest (Failed)
	 99 - NestedPutUTest (Failed)
	110 - BackwardChainerUTest (Failed)
```

**Comment by linas** [comment]
Um, these unit tests never-ever fail for me; they've been passing just fine for years (well two of them are newer than that, but they always pass)  For me, 121 tests total run, and they always pass.

We need a distinct bug report for each, to track down the problems.  VectorAPI might be failing for the same reason that Nil has already reported.

**Comment by AmeBel** [comment]
#### Master branch unit test states
```
cogutil master sha = d789e4c                                          
100% tests passed, 0 tests failed out of 12                       
                                                                  
-----------------------                                           
atomspace master sha = ac76f43                                        
100% tests passed, 0 tests failed out of 120                      
                                                                  
-----------------------                                           
opencog master sha = cc9e331                                          
80% tests passed, 4 tests failed out of 20                        
                                                                  
Total Test time (real) = 225.43 sec                               
                                                                  
The following tests FAILED:                                       
          3 - CogServerUTest (SEGFAULT)                           
          6 - AtomOcTreeUTest (SEGFAULT)                          
          7 - TimeSpaceAtomUTest (SEGFAULT)                       
         11 - PLNRulesUTest (Failed)                              
                                                                  
```
#### Pull request's unit test states
```
cogutil pr 120 sha = 8f7c7e5                                          
100% tests passed, 0 tests failed out of 12                       
                                                                  
-----------------------                                           
atomspace pr 1832 sha= f9809df                                       
atomspace pr 1823 rebased on master sha = 62701cd                                     
100% tests passed, 0 tests failed out of 120                      
                                                                  
-----------------------                                           
opencog master = cc9e331                                          
80% tests passed, 4 tests failed out of 20                        
                                                                  
Total Test time (real) = 225.76 sec                               
                                                                  
The following tests FAILED:                                       
          3 - CogServerUTest (SEGFAULT)                           
          6 - AtomOcTreeUTest (SEGFAULT)                          
          7 - TimeSpaceAtomUTest (SEGFAULT)                       
         11 - PLNRulesUTest (Failed)                              
                                                                  
```

**Comment by vsbogd** [comment]
Raised issue https://github.com/opencog/atomspace/issues/1834 on VectorAPIUTest failure. Looks like it fails on clean build because there are no OpenCog node types which are used in test data.

**Comment by vsbogd** [comment]
Other tests fail because I didn't do ```make install``` before running. Raised separate issue https://github.com/opencog/atomspace/issues/1835 on this

File: cmake/OpenCogFunctions.cmake [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+FUNCTION(PROCESS_MODULE_STRUCTURE FILE_NAME)
+    SET(GUILE_SYMLINK_DIR "${CMAKE_BINARY_DIR}/opencog/scm")
+    SET(GUILE_INSTALL_DIR "${DATADIR}/scm")
+
+    # Create symlinks in build directory mirroring the install path structure.
+    # Also configure for install.
+    IF ("${MODULE_NAME}.scm" STREQUAL "${FILE_NAME}")
+        EXECUTE_PROCESS(
+            COMMAND ${CMAKE_COMMAND} -E make_directory ${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}
+            COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_SOURCE_DIR}/${FILE_NAME}" "${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}/${FILE_NAME}"
```
Comment (vsbogd) [inline cmake/OpenCogFunctions.cmake]: It is not this PR issue but rather an issue that exists in opencog/atomspace master branch.

EXECUTE_PROCESS chaining make_directory and create_symlink actions fails sporadically because directory is not created at the moment when symlinks is being created:
```
failed to create symbolic link '/tmp/test/atomspace-build/opencog/scm//opencog/matrix.scm': No such file or directory
```
Can be fixed by replacing one EXECUTE_PROCESS with two commands by two EXECUTE_PROCESS lines with one command in each:

    IF ("${MODULE_NAME}.scm" STREQUAL "${FILE_NAME}")
        EXECUTE_PROCESS(
            COMMAND ${CMAKE_COMMAND} -E make_directory ${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}
        )
        EXECUTE_PROCESS(
            COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_SOURCE_DIR}/${FILE_NAME}" "${GUILE_SYMLINK_DIR}/${MODULE_FILE_DIR_PATH}/${FILE_NAME}"
        )
        SET(FILE_INSTALL_PATH "${GUILE_INSTALL_DIR}/${MODULE_FILE_DIR_PATH}"
            PARENT_SCOPE
        )
    ELSE()
        EXECUTE_PROCESS(
            COMMAND ${CMAKE_COMMAND} -E make_directory ${GUILE_SYMLINK_DIR}/${MODULE_DIR_PATH}
        )
        EXECUTE_PROCESS(
            COMMAND ${CMAKE_COMMAND} -E create_symlink "${CMAKE_CURRENT_SOURCE_DIR}/${FILE_NAME}" "${GUILE_SYMLINK_DIR}/${MODULE_DIR_PATH}/${FILE_NAME}"
        )
        SET(FILE_INSTALL_PATH "${GUILE_INSTALL_DIR}/${MODULE_DIR_PATH}"
            PARENT_SCOPE
        )
    ENDIF()

Raising it here because it should be fixed either in this PR or after this PR is merged.

**Comment by ngeiswei** [comment]
Testing this is proving more difficult. I could persevere but it seems it has been sufficiently tested by you guy (thanks a lot for your effort!).

I'll wait 24h in case you have something else to say about it and I'll merge.

**Comment by noskill** [comment]
@ngeiswei I moved some opencog cmake files here as well.

**Comment by ngeiswei** [comment]
@vsbogd I've sent you an invite so you have merging rights, to share a bit the burden :-) hopefully nobody has something against that and I trust you'll use your rights wisely.

**Comment by vsbogd** [comment]
Thanks, @ngeiswei, will do my best to do the things right.

**Comment by noskill** [comment]
@vsbogd I removed commit with opencog cmake files..

**Comment by linas** [comment]
> merging rights

Yeah, OK.

**Comment by ngeiswei** [comment]
Thanks @noskill, and thanks @vsbogd for taking care of it.

**Comment by linas** [comment]
FYI, in issue opencog/atomspace#1923, it has been observed that three files were moved from atomspace to cogutils by mistake: `OpenCogAtomTypes.cmake`, `OpenCogFunctions.cmake` and `OpenCogMacros.cmake` I'm encouraging the required pull reqs to move them back to the atomspace, and remove them from cogutils.

**Comment by Dagiopia** [comment]
But that is only useful if the atomspace is installed. It makes sense to have it in the atomspace repo don't it.

**Comment by linas** [comment]
@noskill  Yes, I'm the one who designed it so that other projects use OPENCOG_ADD_ATOM_TYPES.  :-)  So @Dagiopia  is correct, that macro only makes sense if the atomspace is installed; it references files that exist only in the atomspace, and requires glue code that exists only in the atomspace. 

It really did get moved by accident; I just wasn't paying attention when your original pull req went through; and apparently no one else was, either. Accidents happen.

**Comment by linas** [comment]
The point is: @Dagiopia please do go ahead to create a pull req to fix this. It will take a while to get through, as we need a cooling-off period to make sure that the other repos (opencog, as-moses and agi-bio) are not adversely affected. With a bit of luck, this change should be invisible to the other repos.
